File: problem_12/python/problem12.py
#!/usr/bin/python

"""
solution for problem 12 in python.
"""
import math

def get_divisors(number):
    tlist = []
    for x in xrange(2, int(math.sqrt(number))):
        d,r = divmod(number,x)
        if r == 0:
            tlist.append(x)
            tlist.append(d)

    return len([1, number] + tlist)

def triangle_nums():
    iterator = 7
    num = 28

    while True:
        yield num
        iterator += 1
        num += iterator

if __name__ == "__main__":
    tn = triangle_nums()
    for t in tn:
        tl = get_divisors(t)
        if tl > 500:
            print "num: %d\ncount: %d" % (t,tl)
            break

File: problem_12/perl/problem12.pl
#!/usr/bin/perl

use strict;
use warnings;

my $index = 7;
my $total = 28;
my $divisors = 0;

sub divisors
{
    my ($number) = @_;
    my $sq_n = sqrt($number);
    my $i = 1;
    my $t = 0;

    while ($i <= $sq_n)
    {
        if ($number % $i == 0)
        {
            $t += 2;
        }

        $i += 1;
    }

    return $t;
}

while( divisors($total) <= 500)
{
    $index += 1;
    $total += $index;
}

print "$total\n";
